core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +120,146 @@ where
+        assert!(length > 0, "range must not be empty");
+        assert!(length <= self.hashes.len(), "not enough leaves in the tree");
+        let mut right_path = vec![];
+        let root_hash =
+            self.compute_merkle_root_and_path(length - 1, Some(&mut right_path), Some(Side::Right));
+        (root_hash, self.cache.clone(), right_path)
+    }
+
+    /// Adds a raw hash to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push_hash(&mut self, leaf_hash: H256) {
+        self.hashes.push_back(leaf_hash);
+        if self.start_index + self.hashes.len() > self.binary_tree_size {
+            self.binary_tree_size *= 2;
+            if self.cache.len() < tree_depth_by_size(self.binary_tree_size) {
+                self.cache.push(None);
+            }
+        }
+    }
+
+    /// Adds a new leaf to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push(&mut self, leaf: [u8; LEAF_SIZE]) {
+        let leaf_hash = self.hasher.hash_bytes(&leaf);
+        self.push_hash(leaf_hash);
+    }
+
+    /// Caches the rightmost `count` leaves.
+    /// Does not affect the root hash, but makes it impossible to get the paths to the cached leaves.
+    /// # Panics
+    /// Panics if `count` is greater than the number of non-cached leaves in the tree.
+    pub fn trim_start(&mut self, count: usize) {
+        assert!(self.hashes.len() >= count, "not enough leaves to cache");
+        let mut new_cache = vec![];
+        // Cache is a left subset of the path to the first untrimmed leaf.
+        let root = self.compute_merkle_root_and_path(count, Some(&mut new_cache), Some(Side::Left));
+        self.hashes.drain(..count);
+        self.start_index += count;
+        if self.start_index == self.binary_tree_size {
+            // In order to be able to get the root of a completely trimmed tree,
+            // we need to cache it.
+            new_cache.push(Some(root));
+        }
+        self.cache = new_cache;
     }
 
     fn compute_merkle_root_and_path(
-        self,
-        mut index: usize,
-        mut merkle_path: Option<&mut Vec<H256>>,
+        &self,
+        mut end_index: usize,
+        mut end_path: Option<&mut Vec<Option<H256>>>,
+        side: Option<Side>,
     ) -> H256 {
-        assert!(index < self.hashes.len(), "invalid tree leaf index");
-
         let depth = tree_depth_by_size(self.binary_tree_size);
-        if let Some(merkle_path) = merkle_path.as_deref_mut() {
-            merkle_path.reserve(depth);
+        if let Some(right_path) = end_path.as_deref_mut() {
+            right_path.reserve(depth);
+        }
+
+        let mut hashes = self.hashes.clone();
+        let mut absolute_start_index = self.start_index;
+
+        if hashes.is_empty() {
+            hashes.push_back(self.hasher.empty_subtree_hash(0));
         }
```

Comment:
This doesn't look necessary. If there's a cached leaf hash at 0th level, then the 0th level will be constructed from 2 entries correctly. If there's no cached leaf hash at 0th level, then pushing an empty tree hash will compute `self.hasher.empty_subtree_hash(1)`, which would be used on level 1 anyway.